torch/_utils.py

```diff
@@ -774,31 +781,39 @@ def _get_device_index(
     device of the supported runtime platform if :attr:`optional` is ``True``.
     i.e., the current default CUDA device will be returned if CUDA runtime is supported.
     """
+    if isinstance(device, int):
+        return device
     if isinstance(device, str):
         device = torch.device(device)
-    device_idx: Optional[int] = None
     if isinstance(device, torch.device):
+        if allow_cpu and device.type == "cpu":
+            return -1
+        available_device_type = _get_available_device_type()
+        if available_device_type and device.type != available_device_type:
+            raise ValueError(
+                f"Expected a {available_device_type}{' or cpu' if allow_cpu else ''} device, but got: {device}"
+            )
         if not allow_cpu and device.type == "cpu":
             raise ValueError(f"Expected a non cpu device, but got: {device}")
-        device_idx = -1 if device.type == "cpu" else device.index
-    if isinstance(device, int):
-        device_idx = device
-    if device_idx is None:
-        if optional:
-            # The eager API _get_current_device_index uses `lambda` functions which are
-            # not supported in JIT and hence not scriptable. The JIT equivalent API to get
-            # the current device index is `get_current_device_index()` which can
-            # be scripted. We use is_scripting to check the mode we are in and call the
-            # appropriate API.
-            if torch.jit.is_scripting():
-                device_idx = get_current_device_index()
-            else:
-                device_idx = _get_current_device_index()
+        if device.index:
+            return device.index
+    if not torch.jit.is_scripting() and _get_module_device():
+        if isinstance(device, _get_module_device()):
+            return device.idx
```

Comment:
Do we need an assertion here to make sure `device` has `idx` attribute?

Reply:
Not necessary. if `device` has not `idx` attribute, the current device index will be returned by line 812.

Reply:
How can we make sure `_get_module_device()` always returns a class object having `idx` attribute on it?

Reply:
`_get_module_device` ensures to return a `torch.xpu.device` or `torch.cuda.device` or `None`. And there is a `idx` attribute in `torch.xpu.device` or `torch.cuda.device`.